_posts/2021-05-30-spring-elasticache-redis.md

```diff
@@ -0,0 +1,230 @@
+
+    private final UserRepository repository;
+
+    @Cacheable
+    public User getUser(String id){
+        return repository.findById(id).orElseThrow(()->
+                new RuntimeException("No such user found with id"));
+    }
+}
+```
+We have specified the cache name in `@CacheConfig` annotation.
+
+Clusters of the same name need to exist in the ElastiCache:
+
+![ElastiCache Clusters](/assets/img/posts/spring-elasticache-redis/cache-clusters.png)
+
+Technically, Spring Cloud AWS looks for nodes with the same name but since these are **Single Node** clusters the name of the node is the same as the cluster name.
+
+We also need to define cluster names in the `application.yml`. Spring Cloud AWS will use this to scan the ElastiCache to find the clusters:
+
+```yaml
+cloud:
+  aws:
+    elasticache:
+      clusters:
+        -
+          name: product-cache
+          expiration: 100
+        -
+          name: user-cache
+          expiration: 6000
+```
+
+We can also define Time-To-Live for the keys in each cache via `expiration` properties.
+
+If we are using CloudFormation to deploy our application stack in the AWS then one more
+approach exists for us:
+
+Let's say we are using the following CloudFormation to create Clusters:
+
+```yaml
+Parameters:
+
+  InstanceType:
+    Description: Instance Type to use for Cache Cluster
+    Type: String
+    Default: cache.t2.micro
+
+
+Resources:
+
+  ProductCache:
+    Type: AWS::ElastiCache::CacheCluster
+    Properties:
+      ClusterName: product-cache
+      Engine: "Redis"
+      EngineVersion: "6.x"
+      CacheNodeType: !Ref InstanceType
+      NumCacheNodes: 1
+      VpcSecurityGroupIds: ["sg-xxxx"]
+
+  UserCache:
+    Type: AWS::ElastiCache::CacheCluster
+    Properties:
+      ClusterName: user-cache
+      Engine: "Redis"
+      EngineVersion: "6.x"
+      CacheNodeType: !Ref InstanceType
+      NumCacheNodes: 1
+      VpcSecurityGroupIds: [ "sg-xxxx" ]
+```
+
+Then instead of giving cluster names, we only need to provide the stack name. Say the stack name
+is `example-stack`:
+
+```yaml
+cloud:
+  aws:
+    stack:
+      name: example-stack
+```
+
+Spring Cloud AWS retrieves all the cache clusters from our stack and builds `CacheManager` with
+the names of the resources as cache names instead of cluster names. With this, we need to
+update our cache names in our application to resource names instead of cluster names:
+
+```java
+@CacheConfig(cacheNames = "ProductCache")
+public class ProductService {
+    //...
+}
+
+@CacheConfig(cacheNames = "UserCache")
+public class UserService {
+    //...
+}
+```
+
+Make sure to add the following dependency when using the stack name approach:
+
+```groovy
+implementation 'com.amazonaws:aws-java-sdk-cloudformation'
+```
+
+
+## How Does Spring Cloud AWS Configure the `CacheManager`?
+
+In this section, we will dive a bit deeper into the inner workings of the Spring Cloud AWS
+and see how it autoconfigures the cache for us.
```

Comment:
A diagram will help here.

Reply:
This section already has a class diagram, is it not informative? Do you think a flow chart would be better?

Reply:
yes. The sequence of interactions is not clear in the class diagram. The text can also be broken down in bullets.